### PR TITLE
Desktop: Added support for automatic dark/light theme switching

### PR DESCRIPTION
**Pull request by kowalskidev**
Fixed issue: https://github.com/laurent22/joplin/issues/2665
With reference to my previous PR https://github.com/laurent22/joplin/pull/2993

As discussed with here: https://discourse.joplinapp.org/t/night-mode-button/6581/14?u=kowalskidev

> - When auto-detect is enabled
>
>    “Theme” dropdown hidden
>    “Preferred dark theme” visible
>    “Preferred light theme” visible
>
> - When auto-detect disabled
>
>    “Theme” dropdown visible
>    “Preferred dark theme” hidden
>    “Preferred light theme” hidden
>
> While auto-detect is enabled, the theme setting is overwritten with applicable light or dark theme.

I implemented in exact way,

- Added a Auto Detect Checkbox setting enabling it will hide existing `Theme` drop down, Preferred Light/Dark Theme dropdowns will be visible and active Theme will be set according to Preferred Light/Dark Theme option based on system detected theme (Dark theme in my case [As shown in GIF below])

- Added Preferred Light Theme and Preferred Dark Theme dropdowns

- Here is the demo in GIF:
newcommit

@laurent22 Please let me know if any improvements needed.
@PackElend label me please

### CODE REVIEW

File: ElectronClient/gui/ConfigScreen.jsx [hunk 1]
```diff
@@ -316,6 +347,15 @@ class ConfigScreenComponent extends React.Component {
...
+								// preferred dark theme should only work for dark system theme
+								if (key == 'preferredDarkTheme' && this.state.shouldUseDarkColors) {
+									updateSettingValue('theme', event.target.value);
+								} else if (key == 'preferredLightTheme' && !this.state.shouldUseDarkColors) {
+									updateSettingValue('theme', event.target.value);
+								}
+							}
 						}}
 					>
 						{items}
```
Comment (laurent22) [inline ElectronClient/gui/ConfigScreen.jsx:361]: Ideally there should not be any change to this ConfigScreen component. The setting model already provides a "show" property which you can use to hide/show certain settings based on the value of other settings. Also it's nice if the UI updates automatically when you click "auto detect" it's not necessary. In general we want to limit any bespoke code in here and instead enhance, if needed, models/Setting to specify whatever feature is needed.

Comment (kowalskidev) [reply ElectronClient/gui/ConfigScreen.jsx:361]: >Ideally there should not be any change to this ConfigScreen component.

There are multiple reasons to change this component.

- Every time we Enable/Disable `Auto Detect` option we need to detect system theme so to achieve that we need `bridge()` to call `shouldUseDarkColors()` function and `bridge()` is already imported in this component so it's become easy to implement this feature so when we Enable/Disable `Auto Detect` option the state will be change to detected system theme:
```javascript
this.setState({ shouldUseDarkColors: bridge().shouldUseDarkColors() });
```

- Now after enabling `Auto Detect` option when we click on `Apply` button `Theme` should change to `Preferred Light Theme` OR `Preferred Dark Theme`  option based on detected system theme to achieve the same I wrote this line:
```javascript
this.state.shouldUseDarkColors ? changeSetting = 'preferredDarkTheme' : changeSetting = 'preferredLightTheme';
```

- On changing `preferred light/dark theme` setting we also need to change `theme` setting and there's also one validation needed because we have two options visible to the user `Preferred Light Theme` & `Preferred Dark Theme` and only one option should change the `theme` based on detected system theme so to achieve the same I add this lines to the component. 
```javascript
if (key == 'preferredDarkTheme' && this.state.shouldUseDarkColors) {
    updateSettingValue('theme', event.target.value);
} else if (key == 'preferredLightTheme' && !this.state.shouldUseDarkColors) {
    updateSettingValue('theme', event.target.value);
}
```

I couldn't find any other way of implementing the specs without changing ConfigScreen component. If you have any other better way of achieving the same please let me know I will do accordingly

> The setting model already provides a "show" property which you can use to hide/show certain settings based on the value of other settings.

Oopss! I didn't noticed that. Thanks for the hint I'll use that property now.

> Also it's nice if the UI updates automatically when you click "auto detect" it's not necessary. In general we want to limit any bespoke code in here and instead enhance, if needed, models/Setting to specify whatever feature is needed.

I'm little bit confused here, are you saying that I should remove the immediate behavior of setting the active theme to preferred light/dark theme (without clicking on `Apply` button)
as shown in GIF ? Can you please elaborate a little more.

File: ElectronClient/gui/ConfigScreen.jsx [hunk 2]
```diff
@@ -279,8 +292,26 @@ class ConfigScreenComponent extends React.Component {
 			borderRadius: 4,
 		});
 
+		const detectTheme = (autoDetect) => {
+			this.setState({ shouldUseDarkColors: bridge().shouldUseDarkColors() });
+
+			let changeSetting = '';
+			// active theme should change to preferred theme based on system detected theme
+			if (autoDetect) {
```
Comment (kowalskidev) [inline ElectronClient/gui/ConfigScreen.jsx:300]: I'll remove this if() statement as it's not needed now. It was there because of previous spec in which `Preferred Light/Dark` options were always visible

File: ElectronClient/gui/ConfigScreen.jsx [hunk 3]
```diff
@@ -316,6 +347,15 @@ class ConfigScreenComponent extends React.Component {
 						style={selectStyle}
 						onChange={event => {
 							updateSettingValue(key, event.target.value);
+							// if Auto Detect option is ON theme should change according to preferred theme options
+							if (settings['themeAutoDetect']) {
```
Comment (kowalskidev) [inline ElectronClient/gui/ConfigScreen.jsx:351]: I'll remove this if() statement as it's not needed now. It was there because of previous spec in which `Preferred Light/Dark` options were always visible

File: ElectronClient/gui/ConfigScreen.jsx (hunk 1 above)
Comment (laurent22) [reply ElectronClient/gui/ConfigScreen.jsx:361]: > I'm little bit confused here, are you saying that I should remove the immediate behavior of setting the active theme to preferred light/dark theme (without clicking on Apply button) as shown in GIF ? Can you please elaborate a little more.

Yes we should remove this as settings are modified only when the user presses Save. If the theme changes immediately it means something in the app state has changed, and that should not happen.

Regarding your other questions, if you need to modify something in the app as a result of a setting being changed, you can put your code in BaseApplication::generalMiddleware. We can't have setting-specific code in ConfigScreen as it needs to remain generic and work for all settings (I know that there's specific code for sync settings, but ideally that should be moved out too).

Comment (kowalskidev) [reply ElectronClient/gui/ConfigScreen.jsx:361]: > Yes we should remove this as settings are modified only when the user presses Save. If the theme changes immediately it means something in the app state has changed, and that should not happen.

Okay fine I got it..I will remove that behavior.

>Regarding your other questions, if you need to modify something in the app as a result of a setting being changed, you can put your code in BaseApplication::generalMiddleware. We can't have setting-specific code in ConfigScreen as it needs to remain generic and work for all settings (I know that there's specific code for sync settings, but ideally that should be moved out too).

Thanks for the detailed feedback. I understand now and will move the code in ` BaseApplication::generalMiddleware`

File: ElectronClient/gui/ConfigScreen.jsx [hunk 4]
```diff
@@ -281,6 +283,9 @@ class ConfigScreenComponent extends React.Component {
 
 		const updateSettingValue = (key, value) => {
 			// console.info(key + ' = ' + value);
+			if (key == 'themeAutoDetect') {
+				updateSettingValue('shouldUseDarkColors', bridge().shouldUseDarkColors());
```
Comment (kowalskidev) [inline ElectronClient/gui/ConfigScreen.jsx:287]: I tried my best to avoid making changes to this component but I think this line required here to detect system theme because `bridge()` is already imported in this component and also Auto Detect feature is specific to ElectronClient so I think we cannot call `shouldUseDarkColors()` from `BaseApplication.js` as `bridge()` cannot be imported in this file as it's  common for all Clients.
@laurent22  Let me know if I'm wrong somewhere and also if any changes needed.

Comment (laurent22) [reply ElectronClient/gui/ConfigScreen.jsx:285]: But why do you need this shouldUseDarkColors setting value? Can't you read it directly from bridge().shouldUseDarkColors() wherever you need it?

File: ReactNativeClient/lib/BaseApplication.js [hunk 5]
```diff
@@ -529,6 +528,15 @@ class BaseApplication {
...
 			}
 		}
+
+		if (action.type == 'SETTING_UPDATE_ONE' && action.key === 'preferredlightTheme' && !Setting.value('shouldUseDarkColors')) {
+			Setting.setValue('theme', action.value);
+		}
+
+		if (action.type == 'SETTING_UPDATE_ONE' && action.key === 'preferredDarkTheme' && Setting.value('shouldUseDarkColors')) {
+			Setting.setValue('theme', action.value);
+		}
```
Comment (laurent22) [inline ReactNativeClient/lib/BaseApplication.js:538]: The logic above is not right as it's automatically changing the theme even when autodetect is off.

File: ReactNativeClient/lib/models/Setting.js [hunk 6]
```diff
@@ -237,30 +251,60 @@ class Setting extends BaseModel {
...
+				value: Setting.THEME_LIGHT,
+				type: Setting.TYPE_INT,
+				public: true,
+				show: settings => {
+					return settings['themeAutoDetect'];
+				},
+				appTypes: ['desktop'],
+				isEnum: true,
+				label: () => _('Preferred Light Theme'),
+				description: () =>
```
Comment (laurent22) [inline ReactNativeClient/lib/models/Setting.js:287]: Please remove the description

File: ReactNativeClient/lib/models/Setting.js [hunk 7]
```diff
@@ -237,30 +251,60 @@ class Setting extends BaseModel {
...
+				value: Setting.THEME_DARK,
+				type: Setting.TYPE_INT,
+				public: true,
+				show: settings => {
+					return settings['themeAutoDetect'];
 				},
+				appTypes: ['desktop'],
+				isEnum: true,
+				label: () => _('Preferred Dark Theme'),
+				description: () =>
```
Comment (laurent22) [inline ReactNativeClient/lib/models/Setting.js:302]: Please remove the description

File: ElectronClient/gui/ConfigScreen.jsx (hunk 4 above)
Comment (kowalskidev) [reply ElectronClient/gui/ConfigScreen.jsx:285]: >Can't you read it directly from bridge().shouldUseDarkColors() wherever you need it?

I want `bridge().shouldUseDarkColors()` here:
```javascript
if (action.type == 'SETTING_UPDATE_ONE' && action.key === 'preferredDarkTheme' && Setting.value('shouldUseDarkColors')) {
	Setting.setValue('theme', action.value);
}
```
But `bridge()` is not imported in `BaseApplication.js` so created a new setting and set it's value from ConfigScreen.

>But why do you need this shouldUseDarkColors setting value?

We need to check if system detected theme is dark (`bridge().shouldUseDarkColors() == true` ) then only preferred dark theme setting should change `theme` and similarly for light theme.

I don't know if there's any way of calling `shouldUseDarkColors()` function from `BaseApplication.js`

File: ReactNativeClient/lib/BaseApplication.js (hunk 5 above)
Comment (kowalskidev) [reply ReactNativeClient/lib/BaseApplication.js:538]: Yes I intentionally didn't checked for AutoDetect option because these drop downs will only be visible when AutoDetect is ON hence one cannot change this settings from GUI without enabling AutoDetect option. If you still want to add the logic for AutoDetect option then I'll implement it.

File: ReactNativeClient/lib/models/Setting.js (hunk 6 above)
Comment (kowalskidev) [reply ReactNativeClient/lib/models/Setting.js:287]: Okay Sir.

File: ReactNativeClient/lib/models/Setting.js (hunk 7 above)
Comment (kowalskidev) [reply ReactNativeClient/lib/models/Setting.js:302]: Okay Sir.

**Comment by PackElend** [comment]
@kowalskidev why is it closed?

**Comment by kowalskidev** [comment]
> @kowalskidev why is it closed?

There was lots of changes which is not needed for this commits like blank spaces and package-lock.js etc.. so I reset my branch and will submit new commit and then PR will be open.

**Comment by Rishgod** [comment]
there are better ways to do then resetting your branch, now all your commit history is gone, which could have been helpful In the future for someone to see, whenever you mess up a commit don't reset with the upstream but with your last stable commit
`git log --oneline`
get the hash of your last stable commit
`git reset --hard hashcode `
`git push -f`

now your or would be just like it was before you pushed a wrong commit

whenever you are working on an issue, locally create a branch of it

git checkout -b switchcolor

do changes on this branch it and when done with the changes push it 

work on that and don't touch the master, 
you only need a master to (rebase/merge) your branch you are working on.

**Comment by kowalskidev** [comment]
> ```
> There was lots of changes which is not needed for this commits like blank spaces and package-lock.js etc.. so I reset my branch and will submit new commit and then PR will be open.
> ```
> 
> there are better ways to do that then resetting your branch, now all your commit history is gone, which could have been helpful In the future for someone to see, whenever you mess up a commit don't reset with the upstream but with your last stable commit
> `git log --oneline`
> get the hash of your last stable commit
> `git reset --hard hashcode `
> `git push -f`
> 
> now your or would be just like it was before you pushed a wrong commit
> 
> whenever you are working on an issue, locally create a branch of it
> 
> git checkout -b switchcolor
> 
> do changes on this branch it and when done with the changes push it
> 
> work on that and don't touch the master,
> you only need a master to (rebase/merge) your branch you are working on.

@Rishgod  Thank you so much for your help. I'll take care next time.

File: ReactNativeClient/lib/BaseApplication.js (hunk 5 above)
Comment (laurent22) [reply ReactNativeClient/lib/BaseApplication.js:545]: Yes the logic needs to be correct independently of the UI.

File: ElectronClient/gui/ConfigScreen.jsx (hunk 4 above)
Comment (laurent22) [reply ElectronClient/gui/ConfigScreen.jsx:285]: Perhaps then in app/app.js, which overrides BaseApplication. I believe it has a middleware too, specific to the desktop app. You can import bridge there

Comment (kowalskidev) [reply ElectronClient/gui/ConfigScreen.jsx]: Yaa I got it. Thank you sir. Please review the changes, I hope everything is clear now.

File: ElectronClient/app.js [hunk 8]
```diff
@@ -270,6 +270,20 @@ class Application extends BaseApplication {
...
+			Setting.value('themeAutoDetect')) {
+			Setting.setValue('theme', action.value);
+		}
+
+		if (action.type == 'SETTING_UPDATE_ONE' &&
+			action.key === 'preferredDarkTheme' &&
+			bridge().shouldUseDarkColors() &&
+			Setting.value('themeAutoDetect')) {
+			Setting.setValue('theme', action.value);
+    }
```
Comment (laurent22) [inline ElectronClient/app.js:285]: This would not be run when the `themeAutoDetect` checkbox is changed, but it should. For example, consider the example below:

- My system is in dark theme, and I have auto-detect set to "off", with Joplin light theme selected
- I set auto-detect to "on"
- I press OK

Nothing happens, even though the theme should be set to "Dark" like my system.

Also please create a function `detectTheme()` in this file that returns the right theme depending on the values of `shouldUseDarkColors`, `themeAutoDetect` and `preferredLightTheme`/`preferredDarkTheme`. Then use this function in the reducer.

Also make sure that the theme is set correctly when the app starts. If "autoDetect" is on, there should be a call to "detectTheme" and the theme set to the right one.

Comment (kowalskidev) [reply ElectronClient/app.js:285]: > This would not be run when the themeAutoDetect checkbox is changed, but it should

Ohh I missed this edge case 😕 I'll work on it and test again to make sure it works fine with all edge cases

> Also make sure that the theme is set correctly when the app starts. If "autoDetect" is on, there should be a call to "detectTheme" and the theme set to the right one.

Thank you for your detailed feedback, I'll work on it 👍

Comment (laurent22) [reply ElectronClient/app.js:285]: By the way Electron also emits an event when the system theme has changed. This should be used to update the theme when the app is running. https://www.electronjs.org/docs/api/native-theme

**Comment by laurent22** [comment]
Hi @kowalskidev, for now I will close this pull request but you're welcome to complete it and when you've done so, please leave a comment and I'll re-open. Thanks for your work so far.

**Comment by rickythefox** [comment]
This would be a lovely feature to have. @kowalskidev do you think you'll finish it? Otherwise I might take a shot.

**Comment by kowalskidev** [comment]
@rickythefox Actually I unable to work on any projects because of my college lectures, I would definitely love to work on this feature but unfortunately due to the time constraints I unable to devote my time in Joplin So If you wanna go ahead then sure you can give it a shot, How can I stop someone doing awesome things 😁😁
